### PR TITLE
Add wish list page

### PR DESCRIPTION
**Pull request by barshathakuri**
Add wish list page

### CODE REVIEW

File: app/views/BookDetail/index.tsx [hunk 1]
```diff
@@ -113,6 +114,7 @@ function BookDetail() {
                                         <Button
                                             name="wishlist"
                                             variant="secondary"
+                                            onClick={() => history.push('/wish-list')}
```
Comment (AdityaKhatri) [inline app/views/BookDetail/index.tsx]: use ButtonLikeLink

File: app/views/WishList/index.tsx [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+    url: string;
+}
+
+interface Book {
+    id: number;
+    price: number;
+    authors: Author[];
+    image: Image;
+    title: string;
+    mutate: (id: number) => void;
```
Comment (AdityaKhatri) [inline app/views/WishList/index.tsx]: What does this mutate?

File: app/views/WishList/index.tsx [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+
+function WishListBook(props: Book) {
+    const { id, price, authors, image, title, mutate } = props;
+    const authorsDisplay = React.useMemo(() => (
+        authors?.map((d) => d.name).join(', ')
+    ), [authors]);
+
+    const removeBook = (_id: number) => {
+        mutate(_id);
+    };
```
Comment (AdityaKhatri) [inline app/views/WishList/index.tsx]: use useCallback and remove `_` before id

File: app/views/WishList/index.tsx [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+                    {image?.url ? (
+                        <img
+                            className={styles.image}
+                            src={image.url}
+                            alt={title}
+                        />
+                    ) : (
+                        <div className={styles.noPreview}>
+                            Preview not available
+                        </div>
```
Comment (AdityaKhatri) [inline app/views/WishList/index.tsx]: use Message component

File: app/views/WishList/index.tsx [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+                            onClick={undefined}
+                            variant="secondary"
+                            icons={<FaShoppingCart />}
+                            autoFocus
+                        >
+                            Add to cart
+                        </Button>
+                        <Button
+                            name={undefined}
+                            onClick={() => removeBook(id)}
```
Comment (AdityaKhatri) [inline app/views/WishList/index.tsx]:
```suggestion
                            name={id}
                            onClick={removeBook}
```

File: app/views/WishList/index.tsx [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+        WishListQueryVariables
+    >(WISH_LIST, {
+        variables: { page, pageSize },
+    });
+
+    const deleteBook = (_id: number) => {
+        deleteWishlist({ variables: { id: _id } });
+        refetch();
+    };
+
```
Comment (AdityaKhatri) [inline app/views/WishList/index.tsx]: use useCallback

File: app/views/WishList/index.tsx [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+    useEffect(() => {
+        if (!loading && data?.wishList) {
+            if (data.wishList.page != null && data.wishList.page) {
+                setPage(data.wishList.page);
+            }
+            if (data.wishList.pageSize != null && data.wishList.pageSize) {
+                setPageSize(data.wishList.pageSize);
+            }
+        }
+    }, [loading]);
```
Comment (AdityaKhatri) [inline app/views/WishList/index.tsx]: Do this inside onCompleted

File: app/Base/configs/routes.tsx [hunk 8]
```diff
@@ -58,13 +58,23 @@ const bookDetail = wrap({
...
     visibility: 'is-anything',
 });
+const wishList = wrap({
+    path: '/wish-list',
+    title: 'My Wish List',
+    navbarVisibility: true,
+    component: lazy(() => import('#views/WishList')),
+    componentProps: {
+    },
+    visibility: 'is-anything',
```
Comment (tnagorra) [inline app/Base/configs/routes.tsx]: The visibility should not be `is-anything`

File: app/views/WishList/index.tsx [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+
+    return (
+        <div className={styles.wishList}>
+            {!loading && data?.wishList?.results
+                && (
+                    <>
+                        {
+                            data.wishList.results.map((_b: any) => (
```
Comment (AdityaKhatri) [inline app/views/WishList/index.tsx]: define type

File: app/views/BookDetail/index.tsx [hunk 10]
```diff
@@ -113,6 +114,7 @@ function BookDetail() {
...
                                         >
                                             Add to wishlist
```
Comment (tnagorra) [inline app/views/BookDetail/index.tsx]: Why not just use Link component?

File: app/views/WishList/index.tsx [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+                                <WishListBook
+                                    id={_b.book.id}
+                                    title={_b.book.title}
+                                    image={_b.book.image}
+                                    price={_b.book.price}
+                                    authors={_b.book.authors}
+                                    mutate={deleteBook}
```
Comment (AdityaKhatri) [inline app/views/WishList/index.tsx]: do not use _, I don't think that is necessary

File: app/views/WishList/styles.css [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+    flex-direction: column;
+    margin-right: auto;
+    margin-left: auto;
+
+    .container {
+        display: flex;
+        flex-direction: column;
+        margin-top: var(--dui-spacing-large);
+        border-style: solid;
+        border-color: #d3d3d3;
```
Comment (AdityaKhatri) [inline app/views/WishList/styles.css]: do not use colors directly

**Review by AdityaKhatri — CHANGES_REQUESTED** [review]

File: app/views/WishList/styles.css (hunk 12 above)
Comment (tnagorra) [inline app/views/WishList/styles.css]: We should no hard-code color variables.

File: app/views/WishList/index.tsx [hunk 13]
```diff
@@ -0,0 +1,203 @@
...
+                    isbn
+                    authors {
+                        id
+                        name
+                    }
+                    price
+                    title
+                    language
+                    image {
+                        url
```
Comment (tnagorra) [inline app/views/WishList/index.tsx]: If there's an `id` we should always fetch `id`

File: app/views/WishList/index.tsx (hunk 2 above)
Comment (tnagorra) [inline app/views/WishList/index.tsx]: let's rename `mutate` to a better prop name

File: app/views/WishList/index.tsx (hunk 3 above)
Comment (tnagorra) [inline app/views/WishList/index.tsx]: We do not need this function. Let's directly use mutate function

File: app/views/WishList/index.tsx [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+                            </div>
+                        </div>
+                    </Container>
+                    <div className={styles.wishListButton}>
+                        <Button
+                            name={undefined}
+                            onClick={undefined}
+                            variant="secondary"
+                            icons={<FaShoppingCart />}
+                            autoFocus
```
Comment (tnagorra) [inline app/views/WishList/index.tsx]: Any reason to use autoFocus here?

File: app/views/WishList/index.tsx (hunk 5 above)
Comment (tnagorra) [inline app/views/WishList/index.tsx]:
```suggestion
                            name={id}
                            onClick={removeBook}
```

File: app/views/WishList/index.tsx [hunk 15]
```diff
@@ -0,0 +1,203 @@
...
+                        >
+                            Add to cart
+                        </Button>
+                        <Button
+                            name={undefined}
+                            onClick={() => removeBook(id)}
+                            variant="secondary"
+                            icons={<AiTwotoneDelete />}
+                            autoFocus
```
Comment (tnagorra) [inline app/views/WishList/index.tsx]: Any reason to use autoFocus here?

File: app/views/WishList/index.tsx [hunk 16]
```diff
@@ -0,0 +1,203 @@
...
+    const { data, refetch, loading } = useQuery<
+        WishListQuery,
+        WishListQueryVariables
+    >(WISH_LIST, {
+        variables: { page, pageSize },
+    });
+
+    const deleteBook = (_id: number) => {
+        deleteWishlist({ variables: { id: _id } });
+        refetch();
```
Comment (tnagorra) [inline app/views/WishList/index.tsx]: deleteWishlist is a async function, so refetching should be done only on onCompeleted of REMOVE_WISH_LIST mutation.

File: app/views/WishList/index.tsx [hunk 17]
```diff
@@ -0,0 +1,203 @@
...
+                </div>
+            </div>
+        </>
+    );
+}
+
+function WishList() {
+    const [page, setPage] = useState<number>(1);
+    const [pageSize, setPageSize] = useState<number>(10);
+    const [deleteWishlist] = useMutation(REMOVE_WISH_LIST);
```
Comment (tnagorra) [inline app/views/WishList/index.tsx]: Please define typings for this mutation

File: app/views/WishList/index.tsx [hunk 18]
```diff
@@ -0,0 +1,203 @@
...
+
+    const { data, refetch, loading } = useQuery<
+        WishListQuery,
+        WishListQueryVariables
+    >(WISH_LIST, {
+        variables: { page, pageSize },
+    });
+
+    const deleteBook = (_id: number) => {
```
Comment (tnagorra) [inline app/views/WishList/index.tsx]: Any reason we are using underscore before `id`?

Please use useCallback on inline functions.

File: app/views/WishList/index.tsx (hunk 7 above)
Comment (tnagorra) [inline app/views/WishList/index.tsx]: Can you please describe what this block does?

File: app/views/WishList/index.tsx [hunk 19]
```diff
@@ -0,0 +1,203 @@
...
+
+    return (
+        <div className={styles.wishList}>
+            {!loading && data?.wishList?.results
+                && (
+                    <>
+                        {
+                            data.wishList.results.map((_b: any) => (
+                                <WishListBook
+                                    id={_b.book.id}
```
Comment (tnagorra) [inline app/views/WishList/index.tsx]: Prop `key` is missing

File: app/views/WishList/index.tsx [hunk 20]
```diff
@@ -0,0 +1,203 @@
...
+            }
+            if (data.wishList.pageSize != null && data.wishList.pageSize) {
+                setPageSize(data.wishList.pageSize);
+            }
+        }
+    }, [loading]);
+
+    return (
+        <div className={styles.wishList}>
+            {!loading && data?.wishList?.results
```
Comment (tnagorra) [inline app/views/WishList/index.tsx]: Please use ListView and pass this pending state to ListView

**Review by tnagorra — CHANGES_REQUESTED** [review]

File: app/views/WishList/index.tsx (hunk 13 above)
Comment (barshathakuri) [reply app/views/WishList/index.tsx]: There is no id for image in API

File: app/views/WishList/index.tsx [hunk 21]
```diff
@@ -64,25 +67,22 @@ interface Book {
...
     const authorsDisplay = React.useMemo(() => (
         authors?.map((d) => d.name).join(', ')
     ), [authors]);
 
-    const removeBook = (_id: number) => {
-        mutate(_id);
-    };
+    const removeBook = useCallback(() => {
+        removeBookFromWishList(wishListId);
+    }, [wishListId]);
```
Comment (AdityaKhatri) [inline app/views/WishList/index.tsx]: add wishListId as name in button and directly set removeBookFromWishList in button's onclick

File: app/views/WishList/index.tsx [hunk 22]
```diff
@@ -152,45 +152,42 @@ function WishListBook(props: Book) {
...
-    }, [loading]);
-
     return (
         <div className={styles.wishList}>
             {!loading && data?.wishList?.results
                 && (
                     <>
                         {
-                            data.wishList.results.map((_b: any) => (
+                            data.wishList.results.map((b: any) => (
```
Comment (AdityaKhatri) [inline app/views/WishList/index.tsx]: type should be better defined.
Use ListView

**Review by AdityaKhatri — CHANGES_REQUESTED** [review]

File: app/views/WishList/index.tsx (hunk 7 above)
Comment (barshathakuri) [reply app/views/WishList/index.tsx]: For pagination to be implemented in the future.

File: app/views/WishList/index.tsx [hunk 23]
```diff
@@ -0,0 +1,244 @@
...
+    }, []);
+
+    const deleteBook = useCallback((id: string) => {
+        deleteWishlist({ variables: { id } }).then(() => {
+            refetch();
+        });
+    }, []);
+
+    const wishes = (!loading && data?.wishList?.results) ? data.wishList.results : [];
+    const wishItemRendererParams = React.useCallback((_, d) => ({
```
Comment (tnagorra) [inline app/views/WishList/index.tsx]: Add typings for d

File: app/views/WishList/index.tsx [hunk 24]
```diff
@@ -0,0 +1,244 @@
...
+        }).catch((e) => console.log(e));
+    }, []);
+
+    const deleteBook = useCallback((id: string) => {
+        deleteWishlist({ variables: { id } }).then(() => {
+            refetch();
+        });
+    }, []);
+
+    const wishes = (!loading && data?.wishList?.results) ? data.wishList.results : [];
```
Comment (tnagorra) [inline app/views/WishList/index.tsx]: We don't need to check for loading state to read results

File: app/views/WishList/index.tsx [hunk 25]
```diff
@@ -0,0 +1,244 @@
...
+        createCartItem({ variables: { id, quantity } }).then(() => {
+            refetch();
+        }).catch((e) => console.log(e));
+    }, []);
+
+    const deleteBook = useCallback((id: string) => {
+        deleteWishlist({ variables: { id } }).then(() => {
+            refetch();
+        });
+    }, []);
```
Comment (tnagorra) [inline app/views/WishList/index.tsx]: Any reason to use promise here?

File: app/views/WishList/index.tsx [hunk 26]
```diff
@@ -0,0 +1,244 @@
...
+        title,
+        image,
+    } = book;
+
+    const [quantity, setQuantity] = useState<number | undefined>(0);
+    const handleQuantityChange = (value: number | undefined) => {
+        setQuantity(value);
+    };
+    const handleAddToCart = useCallback(() => {
+        if (quantity && quantity > 0) {
```
Comment (tnagorra) [inline app/views/WishList/index.tsx]: We should use isDefined to check if quantity is defined

File: app/views/WishList/index.tsx [hunk 27]
```diff
@@ -0,0 +1,244 @@
...
+        price,
+        authors,
+        title,
+        image,
+    } = book;
+
+    const [quantity, setQuantity] = useState<number | undefined>(0);
+    const handleQuantityChange = (value: number | undefined) => {
+        setQuantity(value);
+    };
```
Comment (tnagorra) [inline app/views/WishList/index.tsx]: Please use useCallback

File: app/views/WishList/index.tsx [hunk 28]
```diff
@@ -0,0 +1,244 @@
...
+  }
+`;
+
+type Wish = NonNullable<NonNullable<WishListQuery['wishList']>['results']>[number]
+const wishKeySelector = (w: Wish) => w.id;
+
+interface WishProps {
+    wish: Wish;
+    removeWishList: (id: string) => void;
+    createCart: (id: string, quantity: number) => void;
```
Comment (tnagorra) [inline app/views/WishList/index.tsx]: Let's rename these props to onCreateCart and onRemoveWishList

File: app/views/BookDetail/index.tsx [hunk 29]
```diff
@@ -110,12 +127,13 @@ function BookDetail() {
                                         <Button name="buy">
                                             Buy now
                                         </Button>
-                                        <Button
-                                            name="wishlist"
+                                        <ButtonLikeLink
                                             variant="secondary"
+                                            onClick={addToWishList}
+                                            to="/wish"
```
Comment (tnagorra) [inline app/views/BookDetail/index.tsx]: Let's not use link directly

File: app/views/BookDetail/index.tsx [hunk 30]
```diff
@@ -55,6 +67,11 @@ function BookDetail() {
         variables: { id: id ?? '' },
     });
 
+    const [createWishList] = useMutation(CREATE_WISH_LIST);
+    const addToWishList = () => {
+        createWishList({ variables: { id } });
+    };
```
Comment (tnagorra) [inline app/views/BookDetail/index.tsx]: Please use useCallback

**Review by tnagorra — CHANGES_REQUESTED** [review]

File: app/views/BookDetail/index.tsx [hunk 31]
```diff
@@ -20,6 +22,8 @@ import {
 } from '#generated/types';
 
 import styles from './styles.css';
+import SmartButtonLikeLink from '#base/components/SmartButtonLikeLink';
+import routes from '#base/configs/routes';
```
Comment (tnagorra) [inline app/views/BookDetail/index.tsx:28]: Let's move absolute import before relative imports

File: app/views/WishList/index.tsx [hunk 32]
```diff
@@ -0,0 +1,294 @@
...
+        wish: d,
+        onRemoveWishList: deleteBook,
+        onCreateCart: addToCart,
+    }), []);
+
+    return (
+        <div className={styles.wishList}>
+            <Container
+                className={styles.featuredBooksSection}
+                heading="My Wishlist"
```
Comment (tnagorra) [inline app/views/WishList/index.tsx:279]: Inconsistent name for "Wishlist".
Let's make sure we are using either "Wish list" or "Wishlist"

File: app/views/WishList/index.tsx [hunk 33]
```diff
@@ -0,0 +1,294 @@
...
+    const [page, setPage] = useState<number>(1);
+    const [pageSize, setPageSize] = useState<number>(10);
+    const alert = useAlert();
+    const [
+        deleteWishlist,
+    ] = useMutation<RemoveWishListMutation, RemoveWishListMutationVariables>(
+        REMOVE_WISH_LIST,
+        {
+            onCompleted: (response) => {
+                refetch();
```
Comment (tnagorra) [inline app/views/WishList/index.tsx:214]: Let's refetch only if mutation is successful

File: app/views/WishList/index.tsx [hunk 34]
```diff
@@ -0,0 +1,294 @@
...
+          isbn
+          authors {
+            id
+            name
+          }
+          price
+          title
+          language
+          image {
+            url
```
Comment (tnagorra) [inline app/views/WishList/index.tsx]: Let's fetch id if image has id

File: app/views/WishList/index.tsx [hunk 35]
```diff
@@ -0,0 +1,294 @@
...
+        book,
+    } = wish;
+
+    const {
+        id: bookId,
+        price,
+        authors,
+        title,
+        image,
+    } = book;
```
Comment (tnagorra) [inline app/views/WishList/index.tsx:104]: Any reason we are de-structuring book?

File: app/views/WishList/index.tsx [hunk 36]
```diff
@@ -0,0 +1,294 @@
...
+
+    const [
+        quantity,
+        setQuantity,
+    ] = useState<number | undefined>(0);
+
+    const handleQuantityChange = useCallback((value: number | undefined) => {
+        setQuantity(value);
+    }, []);
```
Comment (tnagorra) [inline app/views/WishList/index.tsx:113]: We don't need to create separate function handleQuantityChange

File: app/views/WishList/index.tsx [hunk 37]
```diff
@@ -0,0 +1,294 @@
...
+            onCreateCart(bookId, quantity);
+        }
+    }, [quantity]);
+
+    const authorsDisplay = React.useMemo(() => (
+        authors?.map((d) => d.name).join(', ')
+    ), [authors]);
+
+    return (
+        <div className={styles.container} key={id}>
```
Comment (tnagorra) [inline app/views/WishList/index.tsx:126]: Any reason to add key here?

**Review by tnagorra — CHANGES_REQUESTED** [review]

File: app/views/WishList/index.tsx (hunk 34 above)
Comment (barshathakuri) [reply app/views/WishList/index.tsx]: there is no id for image

**Review by AdityaKhatri — APPROVED** [review]

**Pull request merged**